app/src/main/java/ai/elimu/vitabu/ui/storybook/ChapterFragment.java

```diff
@@ -240,48 +200,87 @@ public void playAudio(final TextView textView, final String chapterText, final A
 
-                    // Remove highlighting of the last spoken word
-                    textView.setText(chapterText);
+            @Override
+            public void onStart(String utteranceId) {
+                Log.i(getClass().getName(), "onStart");
+            }
 
-                    if (audioListener != null) {
-                        audioListener.onAudioDone();
+            @Override
+            public void onRangeStart(String utteranceId, int start, int end, int frame) {
+                Log.i(getClass().getName(), "onRangeStart");
+                super.onRangeStart(utteranceId, start, end, frame);
+
+                Log.i(getClass().getName(), "utteranceId: " + utteranceId + ", start: " + start + ", end: " + end);
+                View itemView;
+
+                // Highlight the word being spoken
+                if (wordPosition[0] > -1) {
+                    itemView = layoutManager.findViewByPosition(wordPosition[0]);
+                    if (itemView != null) {
+                        itemView.setBackground(ContextCompat.getDrawable(getContext(), R.drawable.bg_word_selector));
                     }
                 }
 
-                @Override
-                public void onError(String utteranceId) {
-                    Log.i(getClass().getName(), "onError");
+                wordPosition[0]++;
+                itemView = layoutManager.findViewByPosition(wordPosition[0]);
+                if (chapterRecyclerView.getAdapter().getItemViewType(wordPosition[0]) == WordAdapter.NEW_PARAGRAPH_TYPE) {
+                    wordPosition[0]++;
+                } else if (itemView != null && ((TextView)itemView.findViewById(R.id.word_text)).getText().length() == 0) {
+                    wordPosition[0]++;
                 }
-            });
 
-            Log.i(getClass().getName(), "chapterText: \"" + chapterText + "\"");
-            tts.speak(chapterText, TextToSpeech.QUEUE_FLUSH, null, "0");
-        }
+                scrollToWordIfNotVisible(wordPosition[0]);
+                itemView = layoutManager.findViewByPosition(wordPosition[0]);
+                if (itemView != null) {
+                    itemView.setBackgroundColor(ContextCompat.getColor(getContext(), R.color.colorAccent));
+                }
+            }
+
+            @Override
+            public void onDone(String utteranceId) {
+                Log.i(getClass().getName(), "onDone");
+
+                // Remove highlighting of the last spoken word
+                View itemView = layoutManager.findViewByPosition(wordPosition[0]);
+                if (itemView != null) {
+                    itemView.setBackground(ContextCompat.getDrawable(getContext(), R.drawable.bg_word_selector));
+                }
+            }
+
+            @Override
+            public void onError(String utteranceId) {
+                Log.i(getClass().getName(), "onError");
+            }
+
+            private void scrollToWordIfNotVisible(final int position) {
```

Comment:
Thank you for adding this improvement with automatic scrolling. It makes the app much more user-friendly 🙂